### ISSUE TITLE
Get rid of special scalar arithmetic.

### ISSUE THREAD

**Issue by charris** [numpy/numpy#7002]
Numpy umath has a special scalar arithmetic functions that are about 10x faster than the equivalent ufuncs.
```
In [1]: a = np.float64(1)

In [2]: timeit a*a
10000000 loops, best of 3: 69.5 ns per loop

In [3]: timeit np.multiply(a, a)
1000000 loops, best of 3: 722 ns per loop
```
I contend that in large programs this improvement in execution time is not worth the complexity and maintenance overhead. It is unlikely scalar-scalar arithmetic is a significant factor in execution time of  large programs.  Therefore I propose to use ufuncs for all of the scalar-scalar arithmetic. Doing so would also extend future benefits of the `__numpy_ufunc__` attribute to scalars with minimum extra effort.

Thoughts?

**Comment by cournape** [comment]
I agree in principle, but it would be nice to get the original authors to comment on the rationale for the duplication.

**Comment by seberg** [comment]
IIRC the original authors, or at least travis used to call the huge duplication the biggest mistake. It is understandable though....
There are some things where this would have a huge impact though. The example I can think of easily is solving ODEs, which are typically forced to scalars. Of course they are slow in python, but a 10x slowdown could make it slow even for trivial ODEs.

That said, if we recognize the scalar as a 0-d (contiguous) array (which it is) and avoid the array creation overhead, I bet the difference is almost gone.

**Comment by mhvk** [comment]
For scalars, perhaps the recommendation should be to use python objects and `math` (same speed as scalar math, much faster for `sin`, `cos`, etc., since it avoids the `ufunc` overhead).

It would be interesting to see whether recognizing scalars as 0-d arrays would indeed speed up matters; that seems something that could be done before deprecating scalarmath.

More generally: is there really a reason to keep the scalars at all? Could one not use array scalars everywhere? Currently, it is quite annoying to keep subclasses in place when a scalar output might result.

**Comment by charris** [comment]
@mhvk What we would *like* to do is get rid of scalars and use scalar arrays. But getting there from here is a tortuous project. There are ways we could speed up scalar arrays, having the contents for the usual types in the array, for instance, and we could probably speed up the ufunc machinery more for that special case.

**Comment by mhvk** [comment]
Great! I don't know that I can help directly, but will try (if only while thinking more about `__numpy_ufunc__`).

**Comment by rgommers** [comment]
> What we would like to do is get rid of scalars and use scalar arrays.

That doesn't sound quite right. Numpy scalars may have a maintenance overhead, but they're conceptually much nicer than 0-D arrays for users. An array that you can't index with an integer is pretty horrible actually for the unsuspecting user.

**Comment by njsmith** [comment]
You can't index scalars with an integer either..? Conceptually they're the
same things, and in fact the scalars go to elaborate lengths to give almost
the same api as arrays.

**Comment by rgommers** [comment]
> You can't index scalars with an integer either..? 

Of course, but that's not the point.  Given this:

    In [42]: x1
    Out[42]: array(1.5)

    In [43]: x2
    Out[43]: 1.5

a user is likely to try to index ``x1`` (it says it's an array after all), while he's quite unlikely to try to index a scalar.

**Comment by njsmith** [comment]
Ah, sure, I agree. But now this is a user interface issue that has lots of
possible solutions. Maybe we should change the repr of 0d arrays? Maybe we
should just improve the error messages and documentation so that users
aren't confused when things go wrong? (It's already the case that they'll
encounter 0d arrays in various contexts, so maybe they'll be happier in the
long run if forced to learn this at the same time they learn about arrays
and indexing in the first place.) Maybe we can improve the ergonomics of 0d
arrays somehow?
.
(As a side comment, the fact that float64 and float have identical reprs is
*very* confusing; I suspect 90%+ of numpy users have confused and incorrect
ideas about how this actually works, and just don't realize it.)
.
I'm much more worried about stuff like isinstance(np.float64(0), float) ==
True that will be very difficult or impossible to keep working :-(. I guess
py3 broke isinstance(np.int_(0), int) and while it's been painful -- in
particular the way it breaks json.dump -- it hasn't been as catastrophic as
I might have expected.

**Comment by seberg** [comment]
I will agree fully with trying to remove duplicate code, etc. but I still don't buy the removal of scalars as a separate class (or instance of dtype). My reasons:
  * Without scalar conversion (such as in indexing, but probably others), results would be _views_ unless you do an explicit copy.
  * Arrays are normally mutable, scalars are not, that would seem like a rather big change in semantics.
  * This also leads to hashability. Both could be mitigated with read-only arrays, but then `scalar_array += 1` is problematic. (other cases like `scalar_array += arbitrary_object` may be tricky).
 * We register arrays with `sequence` ABC and scalars with `numbers` ABC. We lie about the sequence for 0-d arrays. With scalars it could be worse. The other way around registering N-D arrays with numbers seems dangerous.
  * I am unclear about `object` arrays. You cannot have an 0-D object array act like the object itself.

I still tend to think that ideally instances of dtypes would be scalars. But that is besides the point. When it comes to code duplication, we should avoid it obviously. But about removing scalars from our python API, I am very doubtful and I think we need to think about it more carefully.

**Comment by njsmith** [comment]
I don't think we should worry about copying or mutability or whatever... the reason scalars have those properties now is that there's some code that explicitly makes it happen; that code can continue to exist either way. (And indeed would have to, because yeah we definitely cannot start returning views or mutable objects from scalar indexing.)

I feel like there are two key topics for discussion:
* The duplication of all the arithmetic code between scalars and ufuncs. Probably this should be treated as an independent issue, b/c fixing scalars to call ufuncs would have no user-visible API/ABI consequences at all! (Except for speed and the buggy places where ufuncs and scalar arithmetic implement different operations, e.g. scalar integer arithmetic does overflow checking and ufunc integer arithmetic does not).
* The existence of top-level classes for all the different scalar types. I really think it would be good to get rid of these somehow if we can, because they're a mess. There's tons of code just to set up all the classes (and in a perfect world where we removed code duplication, they'd all basically be identical wrappers around something like an ndarray internally, just allocating new `PyTypeObject`s so as to get different results from `isinstance` checks), they don't play well with extensible dtypes, they lead to the really weird issubdtype system, etc. This is a change that would require very careful work to pull off, though, if it's even possible.

**Comment by seberg** [comment]
All I wanted to say, is that while implementation wise one could maybe make it even identical to ndarray, I believe it needs to be an instance of a dedicated class on the python side. I had, probably falsely, the impression that more was being put on the table than those two points.

**Comment by ygrange** [comment]
Did this happen in the end? I was compiling some code that was actively looking fir the scalarmath library (in a custom cmake script) and it failed for numpy 1.10.4. I was wondering whether the disappearance of numpy.core.scalarmath was intentional.

**Comment by njsmith** [comment]
@ygrange: not sure what that is, but it's unrelated to this issue. Maybe open a new one?
